tests/solana_tests/primitives.rs

```diff
@@ -599,6 +604,363 @@ fn truncate_uint(n: &mut U256, width: usize) {
+                ],
+                &[],
+                None,
+            );
+            let res = first_operand_rand * second_operand_rand;
+
+            assert_eq!(
+                return_value,
+                vec![ethabi::Token::Uint(U256::from_big_endian(
+                    &res.to_bytes_be()
+                ))]
+            );
+        }
+    }
+}
+
+#[test]
+fn test_overflow_detect_signed() {
+    let mut rng = rand::thread_rng();
+    // For bit sizes from 8..64, LLVM has intrinsic functions for multiplication with overflow. Testing starts from int types of 72 and up. There is no need to test intrinsic LLVM functions.
+    for width in (72..=256).step_by(8) {
+        let src = r#"
+        contract test {
+            function mul(intN a, intN b) public returns (intN) {
+                return a * b;
+            }
+        }"#
+        .replace("intN", &format!("int{}", width));
+        let mut contract = build_solidity(&src);
+
+        contract.constructor("test", &[]);
+
+        // The max value held in signed N bits is - 2^(N-1)..2^(N-1)-1 .Generate a value that will overflow this range:
+        let limit = BigInt::from(2_u32).pow(width - 1).add(1_u32);
+
+        // Divide Limit by a random number
+        let first_operand_rand = rng.gen_bigint((width - 1).into()).sub(1_u32);
+
+        // Calculate a number that when multiplied by first_operand_rand, the result will overflow N bits
+        let mut second_operand_rand = limit / &first_operand_rand;
+
+        if let Sign::Minus = second_operand_rand.sign() {
+            // Decrease by 1 if negative, this is to make sure the result will overflow
+            second_operand_rand = second_operand_rand.sub(1);
+        } else {
+            // Increase by 1 if psotive
+            second_operand_rand = second_operand_rand.add(1);
+        }
+
+        let res = contract.function_must_fail(
+            "mul",
+            &[
+                ethabi::Token::Int(bigint_to_eth(
+                    &first_operand_rand,
+                    width.try_into().unwrap(),
+                )),
+                ethabi::Token::Int(bigint_to_eth(
+                    &second_operand_rand,
+                    width.try_into().unwrap(),
+                )),
+            ],
+            &[],
+            None,
+        );
+
+        assert_ne!(res, Ok(0));
+    }
+}
+
+#[test]
+fn test_overflow_detect_unsigned() {
+    let mut rng = rand::thread_rng();
+    for width in (72..=256).step_by(8) {
+        let src = r#"
+        contract test {
+            function mul(uintN a, uintN b) public returns (uintN) {
+                return a * b;
+            }
+        }"#
+        .replace("intN", &format!("int{}", width));
+        let mut contract = build_solidity(&src);
+
+        contract.constructor("test", &[]);
+
+        for _ in 0..10 {
+            // N bits can hold the range 0..(2^N)-1 . Generate a value that overflows N bits
```

Comment:
```suggestion
            // N bits can hold the range [0, (2^N)-1]. Generate a value that overflows N bits
```
Mathematical notation for range.